### DISCUSSION TITLE
Not able to install framework (Windows 10, Node v20.11.1)

### DISCUSSION THREAD

**Discussion by EE2dev** [observablehq/framework#1540] [Q&A] [answered]
This is my environment:
- Windows 10 Enterprise
- Node v20.11.1

I am not able to set up framework on my work computer.
This is my error message when I run the Getting started example: 

```
> dev
> observable preview

Observable Framework v1.9.0
↳ http://127.0.0.1:3000/

GET /
npm:@observablehq/plot → TypeError: fetch failed
    at node:internal/deps/undici/undici:12345:11
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///C:/AZ_DATEN/_Allgemein/observable/framework/hello4/node_modules/@observablehq/framework/dist/npm.js:194:22 {
  cause: Error: getaddrinfo ENOTFOUND data.jsdelivr.com
      at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:118:26) {
    errno: -3008,
    code: 'ENOTFOUND',
    hostname: 'data.jsdelivr.com'
  }
}
TypeError: fetch failed
    at node:internal/deps/undici/undici:12345:11
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///C:/AZ_DATEN/_Allgemein/observable/framework/hello4/node_modules/@observablehq/framework/dist/npm.js:194:22 {
  cause: Error: getaddrinfo ENOTFOUND data.jsdelivr.com
      at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:118:26) {
    errno: -3008,
    code: 'ENOTFOUND',
    syscall: 'getaddrinfo',
    hostname: 'data.jsdelivr.com'
  }
}
```

**Comment by mbostock** [comment, marked as answer]
Probably because of a corporate proxy. Please upvote https://github.com/observablehq/framework/issues/1168 and see the issue for a workaround.

**Reply by EE2dev** [reply]
Thanks - that's what I suspected. Please advise, if there is any helpfull output I can post.

**Reply by mbostock** [reply]
The workaround is here: https://github.com/observablehq/framework/issues/1168#issuecomment-2028238020

**Reply by EE2dev** [reply]
I did add the first workaround option.
The main page loads correctly as does the example report.
However, when the Example dashboard does not. 

What I did:

1) adding (
```
import {ProxyAgent, setGlobalDispatcher} from "undici";

const proxyAgent = new ProxyAgent({uri: process.env.https_proxy});
setGlobalDispatcher(proxyAgent);
```
to observablehq.config.js.

--> The main page loads correctly as does the example report.
2) However, when I choose Example dashboard. the following error is thrown:
```
socket open /_observablehq
↑ {
  type: 'hello',
  path: '/example-dashboard',
  hash: '81853da6a2f576d4e72f6d5e29acf5887c3bed6ebac9bf9f868d7bec97d43e42'
}
GET /_file/data/launches.csv?sha=a09f289a96e2e16631ca59a3d40fbdd76fd42a6645f570ed1d8e66506ef33d1a
load src/data/launches.csv.js → [missing] node:internal/deps/undici/undici:12345
    Error.captureStackTrace(err, this);
          ^

TypeError: fetch failed
    at node:internal/deps/undici/undici:12345:11
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async text (file:///C:/AZ_DATEN/_Allgemein/observable/framework/hello4/src/data/launches.csv.js:5:20)
    at async file:///C:/AZ_DATEN/_Allgemein/observable/framework/hello4/src/data/launches.csv.js:33:33 {
  cause: Error: getaddrinfo ENOTFOUND planet4589.org
      at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:118:26) {
    errno: -3008,
    code: 'ENOTFOUND',
    syscall: 'getaddrinfo',
    hostname: 'planet4589.org'
  }
}

Node.js v20.11.1
error in 891ms: loader exited with code 1
Error: loader exited with code 1
    at CommandLoader.exec (file:///C:/AZ_DATEN/_Allgemein/observable/framework/hello4/node_modules/@observablehq/framework/dist/dataloader.js:297:13)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///C:/AZ_DATEN/_Allgemein/observable/framework/hello4/node_modules/@observablehq/framework/dist/dataloader.js:240:11
```

**Reply by EE2dev** [reply]
I can also post it [here](https://github.com/observablehq/framework/issues/1168#issuecomment-2028238020) to have the issues on one place.

**Reply by mbostock** [reply]
It could be that your corporate proxy does not allow connections to planet4589.org.

**Reply by EE2dev** [reply]
From a browser I can load the files:

https://planet4589.org/space/gcat/tsv/tables/lv.tsv
https://planet4589.org/space/gcat/tsv/derived/launchlog.tsv

**Reply by mbostock** [reply]
Ah. The workaround in the `observablehq.config.js` file will only help Framework’s _internal_ requests use your proxy (such as loading JavaScript libraries from jsDelivr). For data loaders, you’ll need similar workarounds to also make sure your data loader respects the `HTTPS_PROXY` environment variable — so in this case you’d need to modify `src/data/launches.csv.js` to also use your proxy if you want that request to work. Node.js doesn’t do this by default, though I see some indication in https://github.com/nodejs/node/issues/8381 that it might be coming in a future release…

**Reply by EE2dev** [reply]
I will try to add the section

```
import {ProxyAgent, setGlobalDispatcher} from "undici";

const proxyAgent = new ProxyAgent({uri: process.env.https_proxy});
setGlobalDispatcher(proxyAgent);
```

to the data loader and reply tomorrow.
Thanks a lot for the immediate support!

**Reply by EE2dev** [reply]
Success! Thanks so much! 👍👍

To put the solution together:

1. Fix Errors **when Framework internal requests uses proxy**

→ Add three lines of code from [#1168](https://github.com/observablehq/framework/issues/1168#issuecomment-2028238020) 
 to the file `observable.config.js` as workaround

2. Fix Errors **when data loader uses proxy**

→ Add the same three lines of code to the data loader (e.g. `./data/launches.csv.js`) as workaround


```
import {ProxyAgent, setGlobalDispatcher} from "undici";

const proxyAgent = new ProxyAgent({uri: process.env.https_proxy});
setGlobalDispatcher(proxyAgent);
```
